ClosedXML/Excel/Patterns/ClosedXMLValueComparer.cs

```diff
@@ -0,0 +1,108 @@
+    {
+        private static ClosedXMLValueComparer _defaultComparer = new ClosedXMLValueComparer(StringComparer.OrdinalIgnoreCase);
+        private readonly StringComparer _stringComparer;
+
+        private ClosedXMLValueComparer()
+            : this(StringComparer.OrdinalIgnoreCase)
+        { }
+
+        private ClosedXMLValueComparer(StringComparer stringComparer)
+        {
+            this._stringComparer = stringComparer;
+        }
+
+        public static ClosedXMLValueComparer DefaultComparer { get { return _defaultComparer; } }
+
+        public int Compare(object x, object y)
+        {
+            // get both values
+            IComparable c1;
+            Expression e1 = null;
+            if (x is Expression)
+            {
+                e1 = x as Expression;
+                c1 = e1.Evaluate() as IComparable;
+            }
+            else
+                c1 = x as IComparable;
+
+            IComparable c2;
+            Expression e2 = null;
+            if (y is Expression)
+            {
+                e2 = y as Expression;
+                c2 = e2.Evaluate() as IComparable;
+            }
+            else
+                c2 = y as IComparable;
+
+            // handle nulls
+            if (c1 == null && c2 == null)
+                return 0;
+            if (c2 == null)
+                return -1;
+            if (c1 == null)
+                return +1;
+
+            // make sure types are the same
+            if (c1.GetType() != c2.GetType())
+            {
+                try
+                {
+                    // We should improve the type conversion from e.g. Double To DateTime
+                    if (c1 is DateTime && e2 != null)
+                        c2 = (DateTime)e2;
+                    else if (c2 is DateTime && e1 != null)
+                        c1 = (DateTime)e1;
+                    else
+                        c2 = Convert.ChangeType(c2, c1.GetType()) as IComparable;
+                }
+                catch (InvalidCastException) { return -1; }
+                catch (FormatException) { return -1; }
+                catch (OverflowException) { return -1; }
+                catch (ArgumentNullException) { return -1; }
+            }
+
+            // String comparisons should be case insensitive
+            if (c1 is string s1 && c2 is string s2)
+                return StringComparer.OrdinalIgnoreCase.Compare(s1, s2);
+            else
+                return c1.CompareTo(c2);
+        }
+
+        public new bool Equals(object x, object y)
+        {
+            if (x == null && y == null)
+                return true;
+
+            if (y == null || x == null)
+                return false;
+
+            if (ReferenceEquals(x, y))
+                return true;
+
+            if (!x.GetType().Equals(x.GetType()))
+                return false;
+
+            if (x is string s1 && y is string s2)
+                return _stringComparer.Equals(s1, s2);
```

Comment:
String comparison also differs: in `Compare()` method it is case-insensitive, here it relies on `_stringComparer`.